api/metrics/src/main/java/io/opentelemetry/api/metrics/DoubleCounter.java

```diff
@@ -5,59 +5,44 @@
 
 package io.opentelemetry.api.metrics;
 
-import io.opentelemetry.api.metrics.common.Labels;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.context.Context;
 import javax.annotation.concurrent.ThreadSafe;
 
-/**
- * Counter is the most common synchronous instrument. This instrument supports an {@link
- * #add(double, Labels)}` function for reporting an increment, and is restricted to non-negative
- * increments. The default aggregation is `Sum`.
- *
- * <p>Example:
- *
- * <pre>{@code
- * class YourClass {
- *   private static final Meter meter = OpenTelemetry.getMeterProvider().get("my_library_name");
- *   private static final DoubleCounter counter =
- *       meter.
- *           .doubleCounterBuilder("allocated_resources")
- *           .setDescription("Total allocated resources")
- *           .setUnit("1")
- *           .build();
- *
- *   // It is recommended that the API user keep references to a Bound Counters.
- *   private static final BoundDoubleCounter someWorkBound =
- *       counter.bind("work_name", "some_work");
- *
- *   void doSomeWork() {
- *      someWorkBound.add(10.2);  // Resources needed for this task.
- *      // Your code here.
- *   }
- * }
- * }</pre>
- */
+/** A counter instrument that records {@code double} values. */
 @ThreadSafe
-public interface DoubleCounter extends SynchronousInstrument<BoundDoubleCounter> {
-
+public interface DoubleCounter extends Counter {
+  /**
+   * Record a value with a set of attributes.
+   *
+   * @param value The increment amount. MUST be non-negative.
+   * @param attributes A set of attributes to associate with the count.
+   * @param context The explicit context to associate with this measurement.
+   */
+  public void add(double value, Attributes attributes, Context context);
```

Comment:
The API generally looks reasonable to me. The main point that I'd like to make sure is considered is making sure attributes that are recorded are decouplable from attributes that are exported. I see that there is a View AttributesProcessor - is it possible to use Views to model semantic conventions? For example this list of Attributes

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/http-metrics.md#attributes

It seems like having instrumentation (via the API) know what attributes go with what metric is too tedious, and somewhat inefficient, at least for request metrics. For example, we have this

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpServerMetrics.java#L97

I think we want to make sure the API supports the ability to have one Attributes be collected to all the metrics for a request, and the SDK takes care of "applying the semantic conventions" by picking the attributes based on this spec. If Views can be used for this, and we imagine HttpServerMetricsView or such, a view for each semantic convention, then it sounds good.

Reply:
Views can be used for this, but I'm guessing it'll be a bit awkward, as views are SDK-only right now.

We should chat, as this entire prototype made me realize we need (efficient) mechanisms for dealing with `Attributes` in aggregate, similar to your use case.   Additionally, I don't know if your use case is well supported (or not) here.   We should talk through various options and see if we wind up with a missing abstraction in the API that could help here, or if it can be solved SDK-side...

Reply:
@jsuereth one thing that I know from a "source", is that you can have very good performance if you aggregate using all the "Attributes" in the critical path, then "reduce the attributes" on the result (aggregation of 5sec for example).